core/src/main/java/net/opentsdb/threadpools/UserAwareThreadPoolExecutor.java

```diff
@@ -286,67 +328,136 @@ public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
   class QCRunnableWrapper implements Runnable {
     final Runnable r;
     final QueryContext qctx;
+    final TSDTask tsdTask;
 
-    public QCRunnableWrapper(Runnable r, QueryContext qctx) {
+    public QCRunnableWrapper(Runnable r, QueryContext qctx, TSDTask tsdTask) {
       this.r = r;
       this.qctx = qctx;
+      this.tsdTask = tsdTask;
     }
 
     @Override
     public void run() {
 
       if (qctx == null || qctx.authState() == null) {
-        r.run();
-        return;
+        if (tsdTask == null) {
+          r.run();
+          return;
+        }
       }
 
-      final String user = getUser(qctx);
 
-      int thresholdPerUser = threadThresholdPerUserCnt;
-      Integer limit = perUserLimitOverride.get(user);
-      if (limit != null) {
-        thresholdPerUser = limit;
-      }
+      if (qctx != null && qctx.isClosed()) {
+        // Query pipeline is closed upstream, drop this task.
+        // If a task is already running, it will complete.
+        if (tsdTask == TSDTask.QUERY_CLOSE) {
 
-      AtomicInteger ai = CURRENT_EXECUTIONS.get(user);
-      if (ai == null) {
-        ai = CURRENT_EXECUTIONS.computeIfAbsent(user, v -> new AtomicInteger(0));
+          AtomicInteger ai = getCurrentTaskExecutions().get(tsdTask.getCounterPart());
+          if (ai != null) {
+            ai.decrementAndGet();
+            LOG.debug("Removed the query from state, current number of parallel queries : {}",
+                ai.get());
+          }
+        }
+        String user = qctx == null ? "" : getUser(qctx);
+        LOG.debug("Query context {} for user {} is closed, dropping the task",
+            qctx.query().toString(), user);
+        return;
       }
 
-      // Check if the task needs to be requeued based on the thresholds.
-      int activeCount = ((ThreadPoolExecutor) threadPool).getActiveCount();
-      if (activeCount > threadThresholdCnt) {
-        if (ai.get() > thresholdPerUser) {
-
-          if (disableScheduling) {
-            LOG.info("User ({}) task would have been rescheduled since the currently executed user "
-                + "tasks are {} out of available {} threads", user, ai.get(), activeCount);
+      if (tsdTask != null && tsdTask == TSDTask.QUERY) {
+        String user = qctx == null ? "" : getUser(qctx);
+        AtomicInteger ai = getCurrentTaskExecutions().get(tsdTask);
+        if (ai == null) {
+          ai = getCurrentTaskExecutions().computeIfAbsent(tsdTask, v -> new AtomicInteger(0));
+        }
+        LOG.debug("Total parallel running queries {}", ai.get());
+        if (ai.get() > maxParallelQueries) {
+          if (disableTaskLevelScheduling) {
+            LOG.info(
+                "User({}) {} would have been rescheduled since the currently executed {} "
+                    + "are {} and maximum concurrent tasks allowed are {}",
+                user, tsdTask.name(), tsdTask.name(), ai.get(), maxParallelQueries);
```

Comment:
Last tweak: we need to go ahead and run it here ```// Increment the counter per task..
          ai.incrementAndGet();

          // Schedule the task
          r.run();``` if the scheduling is disabled. will need that flag in an emergency if things go pear shaped and we need to let queries through.